src/rust/catloop/mod.rs

```diff
@@ -447,42 +469,97 @@ impl CatloopLibOS {
         self.catmem.borrow_mut().free_sgarray(sga)
     }
 
+    /// Inserts a queue token into the scheduler.
     pub fn schedule(&mut self, qt: QToken) -> Result<SchedulerHandle, Fail> {
-        match self.qts.get(&qt) {
-            Some((demi_opcode_t::DEMI_OPC_ACCEPT, _)) | Some((demi_opcode_t::DEMI_OPC_CONNECT, _)) => {
-                match self.scheduler.from_raw_handle(qt.into()) {
-                    Some(handle) => Ok(handle),
-                    None => return Err(Fail::new(libc::EINVAL, "invalid queue token")),
-                }
-            },
-            Some((demi_opcode_t::DEMI_OPC_PUSH, _)) | Some((demi_opcode_t::DEMI_OPC_POP, _)) => {
-                self.catmem.borrow_mut().schedule(qt)
-            },
-            _ => return Err(Fail::new(libc::EINVAL, "invalid queue token")),
+        // Check if the queue token came from the Catloop LibOS.
+        if let Some((ref opcode, _)) = self.catloop_qts.get(&qt) {
+            // Check if the queue token concerns an expected operation.
+            if opcode != &demi_opcode_t::DEMI_OPC_ACCEPT && opcode != &demi_opcode_t::DEMI_OPC_CONNECT {
+                let cause: String = format!("unexpected queue token (qt={:?})", qt);
+                error!("schedule(): {:?}", &cause);
+                return Err(Fail::new(libc::EINVAL, &cause));
+            }
+
+            // Insert the queue token into the scheduler.
+            match self.scheduler.from_raw_handle(qt.into()) {
```

Comment:
This code doesn't actually insert the token into the scheduler but resolves it from a queue token to a scheduler handle, correct? The actual insert happens when scheduler.insert() is called in the accept/connect/push/pop.

Reply:
You are right. Actually the function `schedule()` itself should be renamed to something more meaningful. I'll fix the comment in this PR.